docs/design/2022-11-17-ttl-table.md

```diff
@@ -0,0 +1,369 @@
+# Proposal: Support TTL Table
+- Author(s): [lcwangchao](https://github.com/lcwangchao)
+- Tracking Issue: https://github.com/pingcap/tidb/issues/39262
+
+## Table of Contents
+
+<!-- TOC -->
+* [Proposal: Support TTL Table](#proposal--support-ttl-table)
+  * [Table of Contents](#table-of-contents)
+  * [Introduction](#introduction)
+  * [Detailed Design](#detailed-design)
+    * [Syntax](#syntax)
+      * [Create TTL Table](#create-ttl-table)
+      * [Alter a Table with TTL](#alter-a-table-with-ttl)
+      * [Alter to a non-TTL Table](#alter-to-a-non-ttl-table)
+      * [Constraints](#constraints)
+    * [TTL Job Management](#ttl-job-management)
+    * [TTL Job Details](#ttl-job-details)
+      * [Scan Tasks](#scan-tasks)
+      * [Delete Tasks](#delete-tasks)
+    * [New System Variables](#new-system-variables)
+    * [New Metrics](#new-metrics)
+  * [Known Issues](#known-issues)
+  * [Future Works](#future-works)
+  * [Alternative Solutions](#alternative-solutions)
+<!-- TOC -->
+
+## Introduction
+
+The rows in a TTL table will be deleted automatically when they are expired. It is useful for some scenes, for example, delete the expired verification codes. A TTL table will have a column with the type DATE/DATETIME/TIMESTAMP which will be compared with the current time, if the interval between them exceeds some threshold, the corresponding row will be deleted.
+
+## Detailed Design
+
+### Syntax
+
+#### Create TTL Table
+
+The following example shows how to create a TTL table. The column `create_at` is used to specify the creation time of the rows which will be deleted 3 months after that.
+
+```sql
+CREATE TABLE t1 (
+    id int PRIMARY KEY,
+    created_at TIMESTAMP
+) TTL = `created_at` + INTERVAL 3 MONTH;
+```
+
+We can use another `TTL_ENABLE` option to disable/enable the TTL job for the table. For example:
+
+```sql
+CREATE TABLE t1 (
+    id int PRIMARY KEY,
+    created_at TIMESTAMP
+) TTL = `created_at` + INTERVAL 3 MONTH TTL_ENABLE = 'OFF';
+```
+
+The above table will not delete expired rows automatically because `TTL_ENABLE` is set to `OFF`. When the `TTL_ENABLE` is omitted, it uses value `ON` by default.
+
+To make it compatible with mysql, TTL options also support the comment format. For example:
+
+```sql
+CREATE TABLE t1 (
+    id int PRIMARY KEY,
+    created_at TIMESTAMP
+) /*T![ttl] TTL = `created_at` + INTERVAL 3 MONTH TTL_ENABLE = 'OFF'*/;
+```
```

Comment:
Thanks for the doc. 
Just to confirm one thing: this will work even for virtual column?
Imagine if we have an existing table like below with a lot of data, and created_at is unix timestamp in nanosecond.
CREATE TABLE t1 (
    id int PRIMARY KEY,
    created_at long
) 

We dont want to modify existing schema to add a 'DateTime' or 'Date' field, which may cause a lot of existing data to be rewritten.
So I just want to confirm below way can work as expected:
ALTER TABLE t1 ADD column timestamp_ttl bigint AS (FROM_UNIXTIME(created_at/1000000000)) VIRTUAL;
ALTER TABLE t1 ADD TTL = 'timestamp_ttl + INTERVAL 180 DAYS';

If so, can you please help document that (how to enable that on existing tables using long in a light way(without rewriting data))?

Reply:
Yes, it supports generated column with some limitations. Because currently generated column can not be pushed down, so it make take more performance, we'll optimize it in the future. It is documented: https://github.com/pingcap/tidb/pull/39264/files#diff-3199b62f6b171f846c9c0c1670428817d339e0e2ee2b2a35822df8f532c51e58R202

Reply:
Thanks for the quick response!
Just curious how is pushdown involved here, isn't this just some micro batch scan and delete statements?
If there is a performance overhead, is it possible to just support column with long field (unix timestamp)?

Reply:
@lcwangchao 
Could you add a section about "generated column" in this document to elaborate clearly?
We can talk about the compatibility about the generated column and the solution to optimize the generated column calculation.